misc/connection-limits/src/lib.rs

```diff
@@ -233,46 +287,60 @@ impl NetworkBehaviour for Behaviour {
         &mut self,
         connection_id: ConnectionId,
         peer: PeerId,
-        _: &Multiaddr,
-        _: &Multiaddr,
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
     ) -> Result<THandler<Self>, ConnectionDenied> {
         self.pending_inbound_connections.remove(&connection_id);
 
-        check_limit(
-            self.limits.max_established_incoming,
-            self.established_inbound_connections.len(),
-            Kind::EstablishedIncoming,
-        )?;
-        check_limit(
-            self.limits.max_established_per_peer,
-            self.established_per_peer
-                .get(&peer)
-                .map(|connections| connections.len())
-                .unwrap_or(0),
-            Kind::EstablishedPerPeer,
-        )?;
-        check_limit(
-            self.limits.max_established_total,
-            self.established_inbound_connections.len()
-                + self.established_outbound_connections.len(),
-            Kind::EstablishedTotal,
-        )?;
-
+        if !(self.bypass_rules.is_addr_bypassed(local_addr)
+            || self.bypass_rules.is_addr_bypassed(remote_addr)
+            || self.bypass_rules.is_peer_bypassed(&peer))
+        {
+            check_limit(
+                self.limits.max_established_incoming,
+                self.established_inbound_connections.len(),
+                Kind::EstablishedIncoming,
+            )?;
+            check_limit(
+                self.limits.max_established_per_peer,
+                self.established_per_peer
+                    .get(&peer)
+                    .map(|connections| connections.len())
+                    .unwrap_or(0),
+                Kind::EstablishedPerPeer,
+            )?;
+            check_limit(
+                self.limits.max_established_total,
+                self.established_inbound_connections.len()
+                    + self.established_outbound_connections.len(),
+                Kind::EstablishedTotal,
+            )?;
+        }
         Ok(dummy::ConnectionHandler)
     }
 
     fn handle_pending_outbound_connection(
         &mut self,
         connection_id: ConnectionId,
-        _: Option<PeerId>,
-        _: &[Multiaddr],
+        maybe_peer: Option<PeerId>,
+        addresses: &[Multiaddr],
         _: Endpoint,
     ) -> Result<Vec<Multiaddr>, ConnectionDenied> {
-        check_limit(
-            self.limits.max_pending_outgoing,
-            self.pending_outbound_connections.len(),
-            Kind::PendingOutgoing,
-        )?;
+        let mut is_bypassed = false;
+        if let Some(peer) = maybe_peer {
+            is_bypassed = self.bypass_rules.is_peer_bypassed(&peer)
+        }
+        is_bypassed = is_bypassed
+            || addresses
+                .iter()
+                .any(|addr| self.bypass_rules.is_addr_bypassed(addr));
```

Comment:
It can still happen that if there are multiple addresses but not all are in the `BypassRules`, the connection is established to an address that is not bypassed, and then denied in `handle_pending_inbound_connection`, right?
Should we then add a note on `bypass_multiaddr` that all known multiaddresses of the remote must be added in order to guarantee that the bypass will succeed?

Reply:
I'm not sure because the implementation will get complex very soon if we want to implement AND/OR logic. If we want to make it as simple as possible we should only check against PeerID and use a closure for more fine-grain control, what do you think?